<a href="https://colab.research.google.com/github/Juosorioca420/DiscretasII/blob/main/Talleres/Taller01/Taller01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Taller 01
**Teoría de números**

---


####1. ¿Existen enteros $a$ y $b$ tal que $a + b = 544$ y cuyo máximo común divisor es 11?.

Supongamos existen $a,b \in \mathbb{Z}$ cuyo máximo común divisor es $11$ y cuya suma es igual a $544$. Entonces:
\begin{align*}
& 544 = 11n + 11m \quad, \: n,m \in \mathbb{Z} \\
& \frac{544}{11} = (n+m) \quad, \: (n+m) \equiv k \in \mathbb{Z} \\
& \frac{544}{11} = k \\
\end{align*}

Donde $\frac{544}{11} = 49.4545 \notin \mathbb{Z}$, para $ k \in \mathbb{Z}$. Lo que es un absurdo.

De modo que, por reducción al absurdo, NO existen enteros $a$ y $b$ tal que $a+b=544$ y su máximo común divisor sea 11.

####2. Encuentre una regla de divisibilidad para 8 y para 16.


  





  

Podemos escribir un número entero en términos de sus dígitos como:
\begin{align*}
a_n10^n + \cdots + a_110^1+a_0
\end{align*}

> **Regla de divisibilidad para 8:** Un número es divisible por 8 si y solo  si los tres dígitos de menor peso del número forman un múltiplo de 8.
* $5678326\boxed{992}$

  \begin{align*}
  &9(100) + 9(10) + 2 \quad\%  8\\
  &4 + 2 + 2 \quad\% 8\\
  &8 \quad\% 8\\
  &0
  \end{align*}

> **Regla de divisibilidad para 16:** Un número es divisible por 16 si y solo si los cuatro dígitos más a la derecha del número forman un múltiplo de 16.
* $987642\boxed{9984}$

  \begin{align*}
  &9(1000) + 9(100) + 8(10) + 4 \quad\%  16\\
  &1000 + 100 + 0 + 4 \quad\% 16\\
  &40 + 4 + 4 \quad\% 16\\
  &8 + 8  \quad\% 16\\
  &16  \quad\% 16\\
  &0
  \end{align*}

####3. Si $p$ es un número primo y $a^2 \equiv b^2 (\mod p)$, pruebe que $a \equiv \pm b$.

Sean $a,b \in \mathbb{Z}$ cualesquiera, tales que $a^2 \equiv b^2 \pmod{p}$ con $p$ número primo, entonces $p \mid (a^2 - b^2) = (a-b)(a+b)$.

Bajo el **Lema de Euler**: Si $n$ es un número entero y divide a un producto $ab$ y es coprimo con uno de los factores, entonces $n$ divide al otro factor.

En el primer caso se tiene $ (a - b)$ es coprimo con $ p $, puesto que $p$ es primo por hipotesis. A partir del *lema de euler*, se tiene que $p \mid (a+b)$; que por definición de congruencia implica que $a ≡ -b \pmod{p}$.

Analogamente, para $p \mid (a-b)$, pues $p$ tambien es coprimo a $(a+b)$, entonces $a ≡ b \pmod{p}$.

Así pues $a \equiv \pm b (\mod p)$ si $a^2 \equiv b^2 \pmod{p}$.

$\blacksquare$

####4. Encuentre el resto cuando $19^{19}$ es dividido por 5.

Por pequeño teorema de Fermat,  $19^{4} \mod 5 = 1$, entonces:

\begin{align*}
& (19^4)^4 \cdot 19^3 \:\%5 \\
& 1 \cdot (15 + 4)^3 \:\%5 \\
& (0 + 4)^3 \:\%5 \\
& 4^2 \cdot 4 \:\%5 \\
& 16 \cdot 4 \:\%5 \\
& 1 \cdot 4 \:\%5 \\
& 4
\end{align*}

$ 19^{19}  \mod 5 = 4 $

In [3]:
print( (19**19) % 5)

4


#### 5. Encuentre los últimos dos dígitos de $7^{7^{7}}$.

Para hallar el ultimo digito de un entero $n$ basta con hallar $n \pmod{10}$, similarmente para los 2 ultimos digitos hace falta determinar $n \pmod{100}$. Tomemos $n = 7^{7^7}$.

Sabemos que $7^7$ es un producto consecutivo de un numero impar, de modo que $7^7 = 2k  + 1$, con $k = 411771$, entonces:

\begin{align*}
& 7^{7^7} \:\% 100\\
& 7^{2k +1} \:\% 100\\
& (7^{2})^k \cdot 7 \:\% 100
\end{align*}


Donde $7^2 = 49$, siendo todas las potencias pares de $49$ terminadas en $01$ y todas las potencias impares de $49$ terminadas en $49$, es decir:

\begin{align*}
& 49^{2ℓ} \:\:\% 100 \quad = 1\\ 
& 49^{2ℓ+1}\:\:\% 100 \quad = 49
\end{align*}

Siendo $k$ numero impar.

\begin{align*}
& 49 \cdot 7 \:\% 100\\ 
& 343\:\% 100\\ 
& 43
\end{align*}

De modo que los ultimos digitos de $n = 7^{7^7}$ son 43.


In [ ]:
print( ( 7**(7**7) ) % 100 )

43


#### 6. Encuentre $\phi(n)$ para n=35,  n=100, n=51200.

Se implementa la función **Totient de Euler** bajo los criterios:

> Propiedades del Totient de Euler. 
  *   $\phi(1)$ = 1 
  *   $\phi(p^a)$ = $p^a - p^{a-1}$ = $p^{a}\left(1 -\frac{1}{p}\right)$
  * $mcd(m,n) = 1 \:\: ⟶$ $\:\:\phi(m\cdot n)$ = $\phi(m)$$\phi(n)$

Donde, aprovechando la propiedad $iii$, es posible descomponer nuestro $n$ es factores primos, tales que, $n = p_1^ap_2^b \cdots p_s^r$ y $mcd(p_x,p_y) = 1$, de modo que

\begin{align*}
 \phi(n) =& \: \phi(p_1^a) \phi(p_2^b) \cdots \phi(p_s^r)\\
 & p_1^{a}\left(1 -\frac{1}{p_1}\right) \cdots p_s^{r}\left(1 -\frac{1}{p_s}\right), \text{ Dado $ii$ para primos.}\\
 & p_1^{a} \cdots p_s^{r} \left(1 -\frac{1}{p_1}\right) \cdots \left(1 -\frac{1}{p_s}\right)\\
 & n \prod_1^s \left(1 -\frac{1}{p_i}\right)
\end{align*}

In [ ]:
#Funciones Auxiliares
def primo(p):
  for i in range( 2, int(p**0.5) + 1 ):
    if p%i == 0:
      return False
  return True

def mcd(x, y):
  return x if y == 0 else mcd(y, (x % y))


# Totient de Euler
def euler(n):
  # i
  if n == 1 : return 1

  # ii (caso especial para primos)
  if primo(n): return (n-1)

  # ii - iii
  factores = []; val = n
  for i in range( 2, int( n**0.5 ) + 1 ): #Hallar los factores primos de n.
    while n % i == 0:
        factores.append(i)
        n /= i
    if n == 1 : break
  if n > 1: factores.append(n)

  for i in set(factores): #Hallar la Productoria; sin duplicados, no importa el exponente.
    val *= ( 1 - (1 / i) )

  return int(val)

In [ ]:
def main(n):
  val = euler(n)
  print( f'n = {n}\nϕ(n) = {val}\n' )

main(35); main(100); main(51200)

n = 35
ϕ(n) = 24

n = 100
ϕ(n) = 40

n = 51200
ϕ(n) = 20480



#### 7. Usted le pregunta a un robot que quiere comer. El responde “48.879”. Sabiendo que el robot piensa en hexadecimal pero habla en decimal, que le debería dar de comer?

Para realizar la conversión de base 10 a base 16, podemos aplicar el algoritmo de la división, iterando hasta que el cociente obtenido de la operación sea tal que $q < 16$. Luego se toman de 'abajo' hacia 'arriba' partiendo desde este cociente, todos los restos obtenidos de las operaciónes previas, considerando que


> 10 = A

>11 = B

>12 = C

>13 = D

>14 = E

>15 = F




In [ ]:
def dec_hex(n):
  res = []
  q : int; r : int
  while n >= 16 :
    # Division entera para el cociente, 16 mod(16) para hallar el residuo
    q, r = n//16, n % 16 
    res.insert(0, r); n = q
  res.insert(0, q)

  hex = ''; vals = { 10 : 'A', 11: 'B', 12: 'C', 13: 'D', 14: 'E', 15: 'F'}
  for num in res:
    if num in vals.keys(): hex += vals[num]
    else: hex += str(num) 

  return hex

In [ ]:
def main(n):
  hex = dec_hex(n)
  print(f'({n})_10 = ({hex})_16')

main(48879)

(48879)_10 = (BEEF)_16


De modo que el robot penso en **BEEF**, es decir, le deberia dar carne para comer.

#### 8. ¿65.314.638.792 es divisible por 24?.

Para determinar si $65.314.638.792$ es divisible por $24$, se utiliza la regla de divisibilidad por $24$.

Esta regla establece que un número es divisible por $24$ si y solo si es divisible por $8$ y $3$.



> $8 \mid 65.314.638.792$
  \begin{align*}
    & 792 \:\% 8 \\
    & 700+90+2 \:\% 8 \\
    & 60 + 10 + 2 \:\% 8 \\
    & 4 + 2 + 2 \:\% 8 \\
    & 8 \:\% 8 \\
    & 0
  \end{align*}

  > $3 \mid 65.314.638.792$
  \begin{align*}
    & 6+5+3+1+4+6+3+8+7+9+2 \:\% 3 \\
    & 54 \:\% 3 \\
    & 21 + 33 \:\% 3 \\
    & 0 + 0 \:\% 3 \\
    & 0 
  \end{align*}

Por lo tanto, como $65.314.638.792$ es divisible por $8$ y por $3$, entonces es divisible por $24$.

#### 9. Pruebe que $n^p - n$ es divisible por $p$ si $p$ es un número primo.

Sea $n \in \mathbb{Z}$ cualquiera y $p$ un número primo. Se tienen 2 posibles casos, $ p \mid n$ y $ p \nmid n$.



* Sea $ p \mid n$.

  Entonces p tambien divide cualquier escalar de $n$, incluyendo $n^p$, de modo que:
  \begin{align*}
    &n^p - n = p (\overline{k} - k)\\
    &n^p - n = p ℓ \\
    & p \mid (n^p - n)
  \end{align*}

* Sea $ p \nmid n \longrightarrow $ $p$, $n$ coprimos.

  \begin{align*}
   & (n^p - n) = (n\cdot n^{p−1} - n)
  \end{align*}

  Por el pequeño teorema de Fermat, $n^{p-1} \equiv 1 \pmod{p}$, de modo que:


  \begin{align*}
    &n(n^{p-1}) - n \pmod{p} \\
    &n(1) - n \pmod{p}\\
    & 0 \pmod{p}\\
  \end{align*}

  \begin{align*}
  ⟹ (n^p - n) \equiv 0 \pmod{p}
  \end{align*}

  Lo que implica por definición de congruencia y divisibilidad $p \mid (n^p - n)$.


Por lo tanto, en todo caso, $p \mid (n^p - n)$ si $p$ es primo.

$\blacksquare$



#### 10. Encuentre los enteros $x$ y $y$ tal que $314x+159y=1$.

Tengase $a = 314, \:\: b = 159\:$, donde:

In [ ]:
a = 314; b = 159
print(f'mcd(a,b) = {mcd(a,b)}') 

mcd(a,b) = 1


De modo que $x$ y $y$ son equivalentes a los números de Bézout, tales que, $mcd(a,b) = 1 = ax + by$.

De modo que es posible aplicar el Algoritmo para los números de Bézout (Euler Extendido) para hallar dichos coeficientes. Así pues:

In [ ]:
def bezout(a, b):
  neg = (a<0 and b<0)

  v, w = 1, 0 # d = av + bw  ->   b=0; d = a(1) + 0
  v1, w1 = 0, 1 # intercambiar e iniciar en caso base para proxima iteración en caso de existir
  while b != 0:
    q = a//b # División de a y b para obtener el cociente
    v, w,   v1, w1  =  v1, w1,   v - q*v1, w - q*w1 # Actualizar los coeficientes
    a, b = b, a%b  # Actualizar a y b

  return (v,w) if (not neg) else (-v,-w)

In [ ]:
def main(a, b):
  if abs(a) < abs(b): a,b = b,a
  v,w = bezout(a,b)
  d = a*v + b*w

  print( f'Para ({a}, {b}):\nd = {d} = {a}({v}) + {b}({w})\n' )
  
main(a,b)

Para (314, 159):
d = 1 = 314(-40) + 159(79)



De modo que.


> $x = -40$

> $y = 79$



#### 11. Pruebe o controvierta la siguiente afirmación si $a^2 \equiv b^2 (\mod m)$ entonces $a \equiv b (\mod m)$ o $a \equiv -b (\mod m)$

Sea el caso $a = 2, \:\:b = 8,\:\:m = 15 \in \mathbb{Z}$; $\:\: a^2 = 4$ y $b^2 = 64$.

Tenemos $4 \equiv 64 \pmod{15}$, de modo que $a^2 \equiv b^2 (\mod m)$.

Ahora bien, no es cierto que $2 \equiv 8 \pmod{15}$ o $2 \equiv -8 \pmod{15}$. Para $a$, en ambos casos es congruentes a $2$. Para $b$, es congruente a $8$ y $7$ respectivamente.

Por lo tanto, por contraejemplo, la afirmación es falsa.

Así pues, que $a^2 \equiv b^2 \pmod{m} \Rightarrow a \equiv b \pmod{m} \:\: \text{ o }\:\: a \equiv -b \pmod{m}$ no se cumple en todos los casos.

$\blacksquare$

#### 12. Encuentre todos los enteros positivos tales que $1066 \equiv 1776 (\mod m)$.

Aplicando las propiedades de la Aritmetica Modular, podemos transformar la expresión inicial:

\begin{align*}
  &1066 \equiv 1776 \pmod{m}\\
  &1066 - 1066 \equiv 1776 - 1066 \pmod{m}\\
  &0 \equiv 710 \pmod{m}
\end{align*}

A partir de la expresión equivalente resultante, se determina que $m$ es todos auquellos enteros divisores de $710$. 
\begin{align*}
m \in \{x : 710\mid x, \:\: x \in \mathbb{Z^+} \}
\end{align*}

Es posible desarrollar un codigo sencillo para hallar dicho conjunto.

In [ ]:
def divisores(n):
  #Encuentra todos los divisores del número entero n.
  divisores = []
  for i in range(1, int(n**0.5) + 1 ): # De la raiz de n se obtiene la maxima tupla de divisores posible
    if n % i == 0:
      divisores.extend([i, int(n/i)]) # Se añade la tupla (i,j), con j = n/i, donde i*j = n.
  return sorted(divisores)

n = 710
print( f'Divisores para {n}:\n{divisores(n)}' )

Divisores para 710:
[1, 2, 5, 10, 71, 142, 355, 710]


Así pues
\begin{align*}
m \in \{1, 2, 5, 10, 71, 142, 355, 710 \}
\end{align*}

#### 13. Muestre que la diferencia de dos cubos consecutivos nunca es divisible por 5.

Sea $n \in \mathbb{Z}$ cualquiera, se tiene que
\begin{align*}
& (n+1)^3 - n^3 \equiv  \: \ell \\
& ℓ= n^3 + 3n^2 + 3n + 1 - n^3\\
&\ell = 3n(n+1) + 1
\end{align*}

Sabemos que cualquier número entero se puede expresar como una de las formas siguientes: 
$5k \:\:\:$, 
$5k+1\:\:\:$, 
$5k+2\:\:\:$, 
$5k+3\:\:\:$, 
$5k+4\:\:\:$.



1.   $n = 5k$
      \begin{align*}
      & \ell = (3)(5k)(5k +1) +1\\
      & \ell - 1 = 5 \overline{k} \\
      & ⟹ \ell ≡ 1 \pmod{5} \text{ , por definicion de congruencia}
      \end{align*}


Analogamente,    

2.   $n = 5k +1$
      \begin{align*}
      & \ell = 3(5k +1)(5k +2) +1\\
      & \ell = 5(3)(5k^2 + 2k + k) + 6 + 1\\
      & \ell - 7 = 5 \overline{k}\\
      & ⟹ \ell ≡ 7 ≡ 2 \pmod{5}
      \end{align*}

3.   $n = 5k +2$
      \begin{align*}
      & \ell = 3(5k +2)(5k +3) +1\\
      & \ell = 5(3)(5k^2 + 3k + 2k) + 18 + 1\\
      & \ell - 19 = 5 \overline{k}\\
      & ⟹ \ell ≡ 19 ≡ 4 \pmod{5}
      \end{align*}

4.   $n = 5k +3$
      \begin{align*}
      & \ell = 3(5k +3)(5k +4) +1\\
      & \ell = 5(3)(5k^2 + 4k + 3k) + 36 + 1\\
      & \ell - 37 = 5 \overline{k}\\
      & ⟹ \ell ≡ 37 ≡ 2 \pmod{5}
      \end{align*}

5.   $n = 5k +4$
      \begin{align*}
      & \ell = 3(5k +4)(5k +5) +1\\
      & \ell = 5(3)(5k +4)(k +1) +1\\
      & \ell = 5(3)(5k^2 + 5k + 4k) + 60 + 1\\
      & \ell - 61 = 5 \overline{k}\\
      & ⟹ \ell ≡ 61 ≡ 1 \pmod{5}
      \end{align*}

De modo que no existen $n, (n+1) \in \mathbb{Z}$, tales que la diferencia de sus cubos tenga división exacta entre $5$.

$\blacksquare$

      



#### 14. Encuentre un entero positivo $n$ tal que $3^2|n$, $4^2|n+1$, $5^2|n+2$

Es posible determinar el sistema de congruencias siguiente.

\begin{align*}
&n≡0\pmod{9} \\
&n+1≡0\pmod{16} \\
&n+2≡0\pmod{25} \\
\end{align*}
Que es equivalente
\begin{align*}
&n+2≡2\pmod{9} \\
&n+2≡1\pmod{16} \\
&n+2≡0\pmod{25} \\
\end{align*}

Aplicando el teorema del resto chino:

\begin{align*}
&m = 3^24^25^2 = 3600\\
&M_1 = 4^25^2 = 400 \\
&M_2 = 3^25^2 = 225 \\
&M_3 = 3^24^2 = 144\\
\end{align*}

Para determinar los inversos podemos hacer uso del algoritmo de Bézout.

In [ ]:
def main(a, b):
  v,w = bezout(a,b)
  d = a*v + b*w

  print( f'Para {a}v ≡ 1 mod({b}):\nv = {v}  \n' )
  
m = [9, 16, 25]; M = [400, 225, 144]
for i, j in zip(M , m): main(i, j)

Para 400v ≡ 1 mod(9):
v = -2  

Para 225v ≡ 1 mod(16):
v = 1  

Para 144v ≡ 1 mod(25):
v = 4  



De modo que.
\begin{align*}
&v_1 = -2 \\
&v_2 = 1 \\
\end{align*}

Finalmente

\begin{align*}
n+2 &= 2(4^25^2)(-2) + 1(3^25^2)(1) + 0\\
&= -1600 + 225\\
&= -1375 \\
&⟶ \quad -1375 \pmod{m}\\
& \quad \quad \quad (2225 - 3600) \pmod{3600}\\
& \quad \quad \quad (2225 - 0) \pmod{3600}\\
& \quad \quad \quad 2225 \\
&\\
n &= 2225 - 2\\
n &=\mathbf{2223} 
\end{align*}

In [ ]:
n = 2223
print( n % 9 )
print( (n+1) % 16 )
print( (n+2) % 25 )

0
0
0


#### 15. ¿Cuál es el último digito de $7^{355}$?

Para hallar el ultimo digito de un entero $n$, basta con determinar $ n \pmod{10}$. Sea $n = 7^{355}$.

\begin{align*}
&7^{354 +1} \quad\% 10 \quad ,354 = 2n \\
&(7^2)^n \cdot 7 \quad\% 10\\
\end{align*}

Con $n = 177$ impar.

\begin{align*}
&9 \cdot 7 \quad\% 10\\
&63 \quad\% 10\\
&3
\end{align*}

De modo que el ultimo digito de $7^{355}$ es $3$.

In [4]:
print( (7**355) % 10)

3


#### 16. Muestre que $3k+4$ y $4k+5$ no tienen un factor común más grande que 1

Supongase existe $d \in \mathbb{Z}$ tales que  $d > 1  $ y $\quad d \mid 3k+4 $ tanto como $d \mid 4k+5 $. Entonces la división de estos dos factores en $d$ tiene el mismo resto:

\begin{align*}
& 4k+5 \equiv 3k+4 \pmod d \\
& 4k + 5 - 3k - 4 = dn\\
& k + 1 = dn\\
& k  = dn -1\\
\end{align*}

Ahora, como $ d \mid 3k+4 $ por hipotesis, $ 3k+4 = dm $, remplazando:

\begin{align*}
& 3(dn -1) +4 = dm \\
& 4 - 3 = d(m - 3n)\\
& 1 = dℓ\\
\end{align*}

Como $ℓ \in \mathbb{Z}$, $ℓ= 1/d$, donde $d$ debe ser un divisor de 1, es decir, $d=1$. No obstante $d > 1$ por hipotesis, lo que es un absurdo.

Por lo tanto, por reducción al absurdo, $3k+4$ y $4k+5$ no tienen un factor común más grande que 1.

$\blacksquare$
